src/coreclr/jit/helperexpansion.cpp

```diff
@@ -0,0 +1,1043 @@
+    //
+    if (optLoopTableValid && prevBb->bbNatLoopNum != BasicBlock::NOT_IN_LOOP)
+    {
+        maxThreadStaticBlocksCondBB->bbNatLoopNum = prevBb->bbNatLoopNum;
+        threadStaticBlockNullCondBB->bbNatLoopNum = prevBb->bbNatLoopNum;
+        fastPathBb->bbNatLoopNum                  = prevBb->bbNatLoopNum;
+        fallbackBb->bbNatLoopNum                  = prevBb->bbNatLoopNum;
+    }
+
+    // All blocks are expected to be in the same EH region
+    assert(BasicBlock::sameEHRegion(prevBb, block));
+    assert(BasicBlock::sameEHRegion(prevBb, maxThreadStaticBlocksCondBB));
+    assert(BasicBlock::sameEHRegion(prevBb, threadStaticBlockNullCondBB));
+    assert(BasicBlock::sameEHRegion(prevBb, fastPathBb));
+
+    // Extra step: merge prevBb with isInitedBb if possible
+    fgReorderBlocks(/* useProfileData */ false);
+    fgUpdateChangedFlowGraph(FlowGraphUpdates::COMPUTE_BASICS);
+
+    return true;
+}
+
+//------------------------------------------------------------------------------
+// fgExpandHelper: Expand the helper using ExpansionFunction.
+//
+// Returns:
+//    true if there was any helper that was expanded.
+//
+template <bool (Compiler::*ExpansionFunction)(BasicBlock*, Statement*, GenTreeCall*)>
+PhaseStatus Compiler::fgExpandHelper(bool skipRarelyRunBlocks)
+{
+    PhaseStatus result = PhaseStatus::MODIFIED_NOTHING;
+    for (BasicBlock* block = fgFirstBB; block != nullptr; block = block->bbNext)
+    {
+        if (skipRarelyRunBlocks && block->isRunRarely())
+        {
+            // It's just an optimization - don't waste time on rarely executed blocks
+            continue;
+        }
+
+        // Expand and visit the last block again to find more candidates
+        while (fgExpandHelperForBlock<ExpansionFunction>(block))
+        {
+            result = PhaseStatus::MODIFIED_EVERYTHING;
+        }
+    }
+    return result;
+}
+
+//------------------------------------------------------------------------------
+// fgExpandHelperForBlock: Scans through all the statements of the `block` and
+//    invoke `fgExpand` if any of the tree node was a helper call.
+//
+// Arguments:
+//    block    - block to scan for static initializations
+//    fgExpand - function that expands the helper call
+//
+// Returns:
+//    true if a helper was expanded
+//
+template <bool (Compiler::*ExpansionFunction)(BasicBlock*, Statement*, GenTreeCall*)>
+bool Compiler::fgExpandHelperForBlock(BasicBlock* block)
+{
+    for (Statement* const stmt : block->NonPhiStatements())
+    {
+        if ((stmt->GetRootNode()->gtFlags & GTF_CALL) == 0)
+        {
+            // TP: Stmt has no calls - bail out
+            continue;
+        }
+
+        for (GenTree* const tree : stmt->TreeList())
+        {
+            if (!tree->IsHelperCall())
+            {
+                continue;
+            }
+
+            if ((this->*ExpansionFunction)(block, stmt, tree->AsCall()))
+            {
+                return true;
+            }
```

Comment:
It only replaces it when it returns true. When it returns true we'll quite here and re-visit current block again (because after split it will not contain previously visitied statements/trees - all of them will go to new blocks we skip)